map/track.cpp

```diff
@@ -81,14 +43,19 @@ void Track::CacheDataForInteraction()
 
 std::vector<double> Track::GetLengthsImpl() const
 {
-  auto const & line = GetSingleGeometry();
-  std::vector<double> lengths(line.size(), 0.0);
-  for (size_t i = 1; i < line.size(); ++i)
+  auto const & lines = GetData().m_geometry.m_lines;
+  std::vector<double> lengths;
+  lengths.emplace_back(0.0);
+  for (size_t i = 0; i < lines.size(); ++i)
   {
-    auto const & pt1 = line[i - 1].GetPoint();
-    auto const & pt2 = line[i].GetPoint();
-    auto const segmentLength = mercator::DistanceOnEarth(pt1, pt2);
-    lengths[i] = lengths[i - 1] + segmentLength;
+    auto const & line = lines[i];
+    for (size_t j = 1; j < line.size(); ++j)
+    {
+      auto const & pt1 = line[j - 1].GetPoint();
+      auto const & pt2 = line[j].GetPoint();
+      auto const segmentLength = mercator::DistanceOnEarth(pt1, pt2);
+      lengths.emplace_back(lengths.back() + segmentLength);
+    }
```

Comment:
IDK the use-case, but should we make this function the same as in ElevationInfo::ElevationInfo?
Like lenghts.size() == SUM(lines.size()) ?
```
double distance = 0;
for (auto const & line : lines)
{
  lengths.push_back(distance);
  for (size_t j = 1; j < line.size(); ++j)
  {
    ...
    distance += segmentLength;
    lengths.emplace_back(distance);
  }
}
```

We have duplication points, but the logic is consistent with ElevationInfo

Reply:
In this case the distances array  is used to retrieve the proper current point's distance to update the selection
info.
So the duplication may break the logic.
```cpp
void Track::CacheDataForInteraction() const
{
  m_interactionData = InteractionData();
  m_interactionData->m_lengths = GetLengthsImpl(); < ---- here
  m_interactionData->m_limitRect = GetLimitRectImpl();
}

std::vector<double> Track::GetLengthsImpl() const
{
```

```cpp
double Track::GetLengthMetersImpl(kml::MultiGeometry::LineT const & line, size_t ptIdx) const
{
  if (!m_interactionData)
    CacheDataForInteraction();
  CHECK_LESS(ptIdx, m_interactionData->m_lengths.size(), ());
  return m_interactionData->m_lengths[ptIdx];  < ---- here
}

void Track::UpdateSelectionInfo(m2::RectD const & touchRect, TrackSelectionInfo & info) const
{
  if (m_interactionData && !m_interactionData->m_limitRect.IsIntersect(touchRect))
    return;

  for (auto const & line : m_data.m_geometry.m_lines)
  {
    for (size_t i = 0; i + 1 < line.size(); ++i)
    {
      auto pt1 = line[i].GetPoint();
      auto pt2 = line[i + 1].GetPoint();
      if (!m2::Intersect(touchRect, pt1, pt2))
        continue;

      m2::ParametrizedSegment<m2::PointD> seg(pt1, pt2);
      auto const closestPoint = seg.ClosestPointTo(touchRect.Center());
      auto const squaredDist = closestPoint.SquaredLength(touchRect.Center());
      if (squaredDist >= info.m_squareDist)
        continue;

      info.m_squareDist = squaredDist;
      info.m_trackId = m_data.m_id;
      info.m_trackPoint = closestPoint;

      auto const segDistInMeters = mercator::DistanceOnEarth(line[i].GetPoint(), closestPoint);
      info.m_distFromBegM = segDistInMeters + GetLengthMetersImpl(line, I); < --- HERE
    }
  }
}
```

Reply:
In this case the whole logic is invalid here ..

Reply:
I've updated the interaction data and methods to store data not in single vector of doubles, but
```cpp
  using Lengths = std::vector<double>;
  struct InteractionData
  {
    std::vector<Lengths> m_lengths;
    m2::RectD m_limitRect;
  };
```
it fixes the issue in UpdateSelectionInfo
```
info.m_distFromBegM = segDistInMeters + GetLengthMetersImpl(lineIndex, ptIndex);
```